source/socket.c

```diff
@@ -0,0 +1,263 @@
+    }
+
+    // 2. setup vtable based on socket type
+    switch (type) {
+        case AWS_SOCKET_IMPL_POSIX:
+            return aws_socket_init_posix(socket, alloc, options);
+        case AWS_SOCKET_IMPL_WINSOCK:
+            return aws_socket_init_winsock(socket, alloc, options);
+        case AWS_SOCKET_IMPL_APPLE_NETWORK_FRAMEWORK:
+            // Apple Network Framework is not implemented yet. We should not use it yet.
+            AWS_ASSERT(false && "Invalid socket implementation on platform.");
+            return aws_socket_init_apple_nw_socket(socket, alloc, options);
+        default:
+            AWS_ASSERT(false && "Invalid socket implementation on platform.");
+            return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+}
+
+int aws_socket_get_bound_address(const struct aws_socket *socket, struct aws_socket_endpoint *out_address) {
+    if (socket->local_endpoint.address[0] == 0) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p fd=%d: Socket has no local address. Socket must be bound first.",
+            (void *)socket,
+            socket->io_handle.data.fd);
+        return aws_raise_error(AWS_IO_SOCKET_ILLEGAL_OPERATION_FOR_STATE);
+    }
+    *out_address = socket->local_endpoint;
+    return AWS_OP_SUCCESS;
+}
+
+void aws_socket_endpoint_init_local_address_for_test(struct aws_socket_endpoint *endpoint) {
+    (void)endpoint;
+    struct aws_uuid uuid;
+    AWS_FATAL_ASSERT(aws_uuid_init(&uuid) == AWS_OP_SUCCESS);
+    char uuid_str[AWS_UUID_STR_LEN] = {0};
+    struct aws_byte_buf uuid_buf = aws_byte_buf_from_empty_array(uuid_str, sizeof(uuid_str));
+    AWS_FATAL_ASSERT(aws_uuid_to_str(&uuid, &uuid_buf) == AWS_OP_SUCCESS);
+
+#if defined(AWS_ENABLE_KQUEUE) || defined(AWS_ENABLE_EPOLL)
+    snprintf(endpoint->address, sizeof(endpoint->address), "testsock" PRInSTR ".sock", AWS_BYTE_BUF_PRI(uuid_buf));
+    return;
+#endif
+
+#if defined(AWS_ENABLE_IO_COMPLETION_PORTS)
+    snprintf(endpoint->address, sizeof(endpoint->address), "\\\\.\\pipe\\testsock" PRInSTR, AWS_BYTE_BUF_PRI(uuid_buf));
+    return;
+#endif
+}
+
+/**
+ * Return the default socket implementation type. If the return value is `AWS_SOCKET_IMPL_PLATFORM_DEFAULT`, the
+ * function failed to retrieve the default type value.
+ */
+static enum aws_socket_impl_type aws_socket_get_default_impl_type(void) {
+    enum aws_socket_impl_type type = AWS_SOCKET_IMPL_PLATFORM_DEFAULT;
+// override default socket
+#ifdef AWS_USE_APPLE_NETWORK_FRAMEWORK
+    type = AWS_SOCKET_IMPL_APPLE_NETWORK_FRAMEWORK;
+#endif // AWS_USE_APPLE_NETWORK_FRAMEWORK
+    if (type != AWS_SOCKET_IMPL_PLATFORM_DEFAULT) {
+        return type;
+    }
+/**
+ * Ideally we should use the platform definition (e.x.: AWS_OS_APPLE) here, however the platform
+ * definition was declared in aws-c-common. We probably do not want to introduce extra dependency here.
+ */
+#if defined(AWS_ENABLE_KQUEUE) || defined(AWS_ENABLE_EPOLL)
```

Comment:
elif this sequence, otherwise having multiple types enabled leads to multiple return statements with all but one useless